A3-channel-tracing.md

```diff
@@ -0,0 +1,204 @@
+    // channel is connecting
+    GRPC_CHANNEL_CONNECTING = 3;
+    // channel is ready for work
+    GRPC_CHANNEL_READY = 4;
+    // channel has seen a failure but expects to recover
+    GRPC_CHANNEL_TRANSIENT_FAILURE = 5;
+    // channel has seen a failure that it cannot recover from
+    GRPC_CHANNEL_SHUTDOWN = 6;
+  };
+  // state the channel was in when the trace was logged
+  ChannelState state = 5;
+};
+
+// trace data for one channel or subchannel
+message TraceData {
+  // unique id of subchannel associated with this trace data
+  // set to zero if this is a channel's trace data
+  int64 subchannel_id = 1;
+   // tracks total trace nodes seen, since many may be overwritten
+  int64 num_nodes_logged = 2;
+  // time for creation for the channel or subchannel tracer
+  google.protobuf.Timestamp start_time = 3;
+  // all of the data for this tracer
+  repeated TraceNode nodes = 4;
+}
+
+// trace for one channel
+message ChannelTrace {
+  // parent channel data
+  TraceData channel_data = 1;
+  // data for the subchannels of this channel
+  repeated TraceData subchannel_data = 3;
+}
+```
+
+## Rationale
+
+This approach makes it easy to add trace, since each tracer is attached to its respective channel or subchannel. The tracer will also be passed down into the resolver and load balancer to record trace for interesting events that happen there. 
+
+One downside to this approach is the potential for the tracer to bloat memory. For this reason, there will be a new channel argument that sets a limit for how many trace nodes each channel or subchannel will hold. This is described in more detail in the [Garbage Collection](#garbage-collection) section.
+
+## Implementation
+
+### Basic Design
+
+The implementation will be done in the C-core first, with Java and Go to follow. The tracer object will hold a linked list of trace nodes. If a trace node pertains to a subchannel, then it will point to a tracer for that subchannel. tracers are refcounted objects.
+
+```C
+// One node of tracing data
+struct grpc_trace_node {
+  const char* data;
+  grpc_error* error;
+  gpr_timespec time_created;
+  grpc_connectivity_state connectivity_state;
+  grpc_trace_node* next;
+
+  // The tracer object that owns this trace node. This is used to ref and
+  // unref the tracing object as nodes are added or overwritten
+  grpc_channel_tracer* subchannel;
+};
+
+struct grpc_trace_node_list {
+  size_t size;
+  size_t max_size;
+  grpc_trace_node* head_trace;
+  grpc_trace_node* tail_trace;
+};
+
+/* the channel tracing object */
+struct grpc_channel_tracer {
+  gpr_refcount refs;
+  gpr_mu tracer_mu; 
+  int64_t num_nodes_logged;
+  grpc_trace_node_list node_list;
+  gpr_timespec time_created;
+};
+```
+
+### Tracer Functions
+
+The tracer files with expose several calls which with the C-core code can interact with tracers:
+
+```C
+/* Initializes the tracing object with gpr_malloc. The caller has
+   ownership over the returned tracing object */
```

Comment:
Suggest alternative wording:

Creates a new tracer.  The caller owns a reference to the returned tracer.